execution/gethexec/express_lane_service.go

```diff
@@ -0,0 +1,280 @@
+				if err != nil {
+					log.Error("Could not filter auction resolutions", "error", err)
+					continue
+				}
+				for it.Next() {
+					log.Info(
+						"New express lane controller assigned",
+						"round", it.Event.Round,
+						"controller", it.Event.FirstPriceExpressLaneController,
+					)
+					es.Lock()
+					es.roundControl.Add(it.Event.Round, &expressLaneControl{
+						controller: it.Event.FirstPriceExpressLaneController,
+						sequence:   0,
+					})
+					es.Unlock()
+				}
+				fromBlock = toBlock
+			}
+		}
+	})
+}
+
+func (es *expressLaneService) currentRoundHasController() bool {
+	es.Lock()
+	defer es.Unlock()
+	currRound := timeboost.CurrentRound(es.initialTimestamp, es.roundDuration)
+	control, ok := es.roundControl.Get(currRound)
+	if !ok {
+		return false
+	}
+	return control.controller != (common.Address{})
+}
+
+func (es *expressLaneService) isWithinAuctionCloseWindow(arrivalTime time.Time) bool {
+	// Calculate the next round start time
+	elapsedTime := arrivalTime.Sub(es.initialTimestamp)
+	elapsedRounds := elapsedTime / es.roundDuration
+	nextRoundStart := es.initialTimestamp.Add((elapsedRounds + 1) * es.roundDuration)
+	// Calculate the time to the next round
+	timeToNextRound := nextRoundStart.Sub(arrivalTime)
+	// Check if the arrival timestamp is within AUCTION_CLOSING_DURATION of TIME_TO_NEXT_ROUND
+	return timeToNextRound <= es.auctionClosing
+}
+
+func (es *expressLaneService) sequenceExpressLaneSubmission(
+	ctx context.Context,
+	msg *timeboost.ExpressLaneSubmission,
+	publishTxFn func(
+		parentCtx context.Context,
+		tx *types.Transaction,
+		options *arbitrum_types.ConditionalOptions,
+		delay bool,
+	) error,
+) error {
+	es.Lock()
+	defer es.Unlock()
+	control, ok := es.roundControl.Get(msg.Round)
+	if !ok {
+		return timeboost.ErrNoOnchainController
+	}
+	// Check if the submission nonce is too low.
+	if msg.Sequence < control.sequence {
+		return timeboost.ErrSequenceNumberTooLow
+	}
+	// Check if a duplicate submission exists already, and reject if so.
+	if _, exists := es.messagesBySequenceNumber[msg.Sequence]; exists {
+		return timeboost.ErrDuplicateSequenceNumber
+	}
+	// Log an informational warning if the message's sequence number is in the future.
+	if msg.Sequence > control.sequence {
+		log.Warn("Received express lane submission with future sequence number", "sequence", msg.Sequence)
+	}
+	// Put into the the sequence number map.
+	es.messagesBySequenceNumber[msg.Sequence] = msg
+
+	for {
+		// Get the next message in the sequence.
+		nextMsg, exists := es.messagesBySequenceNumber[control.sequence]
+		if !exists {
+			break
+		}
+		if err := publishTxFn(
+			ctx,
+			nextMsg.Transaction,
+			msg.Options,
+			false, /* no delay, as it should go through express lane */
+		); err != nil {
+			// If the tx failed, clear it from the sequence map.
+			delete(es.messagesBySequenceNumber, msg.Sequence)
+			return err
+		}
+		// Increase the global round sequence number.
+		control.sequence += 1
+	}
```

Comment:
Minor because we break, but we might want to check for context timeout here, and the caller of `sequenceExpressLaneSubmission` should set a deadline for when the round ends.